src/ethereum_spec_tools/evm_trace.py

```diff
@@ -1,66 +1,212 @@
-        Add indentation for child evms (depth > 0)
-        """
-        tabs = "\t" * self.depth
-        rep = tabs + self.__repr__()
-        return rep
+    def __init__(self, gas_used: int, output: bytes, has_erred: bool) -> None:
+        self.output = output.hex()
+        self.gasUsed = hex(gas_used)
+        if has_erred:
+            self.error = ""
 
 
-def evm_trace(evm: Any, op: Any) -> None:
+def capture_tx_start(env: Any) -> None:
     """
-    Create a new trace instance before opcode execution
+    Capture the state at the beginning of a transaction.
     """
-    if isinstance(op, bytes):
-        opcode = "0x" + op.hex().lstrip("0")
-    else:
-        opcode = str(op).split(".")[-1]
+    pass
 
-    new_trace = EvmTrace(
-        depth=evm.message.depth,
+
+def capture_tx_end(
+    env: Any, gas_used: int, output: bytes, has_erred: bool
+) -> None:
+    """
+    Capture the state at the end of a transaction.
+    """
+    final_trace = FinalTrace(gas_used, output, has_erred)
+    env.traces.append(final_trace)
+
+
+def capture_precompile_start(evm: Any, address: Any) -> None:
+    """
+    Create a new trace instance before precompile execution.
+    """
+    new_trace = Trace(
         pc=evm.pc,
-        op=opcode,
-        has_erred=False,
-        start_gas=evm.gas_left,
+        op="0x" + address.hex().lstrip("0"),
+        gas=hex(evm.gas_left),
+        gasCost="0x0",
+        memSize=len(evm.memory),
+        stack=[hex(i) for i in evm.stack],
+        depth=evm.message.depth + 1,
         refund=evm.refund_counter,
-        output=evm.output,
-        stack=evm.stack.copy(),
+        opName="0x" + address.hex().lstrip("0"),
+        precompile=True,
+    )
+
+    evm.env.traces.append(new_trace)
+
+
+def capture_precompile_end(evm: Any) -> None:
+    """Capture the state at the end of a precompile execution."""
+    last_trace = evm.env.traces[-1]
+
+    last_trace.gasCostTraced = True
+    last_trace.errorTraced = True
+
+
+def capture_op_start(evm: Any, op: Any) -> None:
+    """
+    Create a new trace instance before opcode execution.
+    """
+    new_trace = Trace(
+        pc=evm.pc,
+        op=op.value,
+        gas=hex(evm.gas_left),
+        gasCost="0x0",
+        memSize=len(evm.memory),
+        stack=[hex(i) for i in evm.stack],
+        depth=evm.message.depth + 1,
+        refund=evm.refund_counter,
+        opName=str(op).split(".")[-1],
+    )
+
+    evm.env.traces.append(new_trace)
+
+
+def capture_op_end(evm: Any) -> None:
+    """Capture the state at the end of an opcode execution."""
+    last_trace = evm.env.traces[-1]
+
+    last_trace.gasCostTraced = True
+    last_trace.errorTraced = True
+
+
+def capture_op_exception(evm: Any) -> None:
+    """Capture the state in case of exceptions."""
+    if any(evm.env.traces):
+        last_trace = evm.env.traces[-1]
+    if (
+        not any(evm.env.traces)
+        or last_trace.errorTraced
+        or last_trace.depth == evm.message.depth
+    ):
```

Comment:
This might be a bit more readable:

```suggestion
    last_trace = None
    if evm.env.traces:
        last_trace = evm.env.traces[-1]

    if (
        not last_trace
        or last_trace.errorTraced
        or last_trace.depth == evm.message.depth
    ):
```

Reply:
Done.